### DISCUSSION TITLE
Collaboration with hierarchical data structures for advanced reports

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [eclipse-birt/birt#997] [Q&A]
[Information sources which provide hierarchical data structures](https://en.wikipedia.org/wiki/Tree_%28data_structure%29 "Description") are also supported to some degree.
The software “Business Intelligence Reporting Tool 4.9” depends on [tabular data for various reports](https://eclipse.github.io/birt-website/docs/t_data-sets/ "Data set mapping").

:thought_balloon: I got the impression that some development challenges trigger corresponding considerations for more convenient and safe data processing.

**Comment by hvbtup** [comment]
The classical use-case for reporting tools like BIRT and others is to create nice output of data from relational databases. That's what BIRT was initially was created for. As a result, the concept of BIRT relies on a data set as tabular data (a rows represents a record, the columns represent the atttributes). 

During the XML hype decades ago, support for reading XML was added. While I have never used XML datasets in production, my impression is that they are not a first-class citizen of the BIRT world.
The current state is that they "work somehow". But there are performance issues when processing big XML data sets.

The performance problem might be caused by the way the XML is processed technically and it might not be an issue for possible other hierarchical data formats like JSON.

But I think there are other issues with hierarchical data formats - not only technically: 

I think many people don't comprehend XPath/XQuery (or JSONPath). This makes extracting details difficult for them. 

And successfully processing herarchical data more or less requires a fixed schema.

I think that in order to improve the handling of hierarchical data formats in BIRT, it would be necessary for someone to come up with a good concept about how this should actually look like and work in an ideal next-generation BIRT.

**Reply by elfring** [reply]
>The current state is that they "work somehow". But there are performance issues when processing big XML data sets.

I am curious if more experiences can be shared for this software area.


>… and it might not be an issue for possible other hierarchical data formats like JSON.

Will any comparisons evolve further for the software run time characteristics? 💭


>I think many people don't comprehend XPath/XQuery (or JSONPath). This makes extracting details difficult for them.

How do you think about to adjust data models on demand according to path dependencies (or corresponding variables)? 🤔


>And successfully processing herarchical data more or less requires a fixed schema.

🔮 Are any dynamic programming approaches helpful and applicable?

**Reply by hvbtup** [reply]
> adjust data models on demand according to path dependencies (or corresponding variables)? 

Please note that you already can use parameters in XML datasets.

Those people who are using XML datasets in production can certainly provide examples and share their experience.

**Comment by TheLaserNinja** [comment]
BIRT was designed to work with anything that is similar to the JDBC data structure, which is why the ODA interface mirrors JDBC so closely. This lends itself well to RDMS. How, as long as you can iterate over it it can be reported on with BIRT. This is how we've done XML and JSON in the past using the Scripted Data Source. But there is nothing about BIRT that would prevent it from reporting off hierarchical structures as long as you can iterate over it in some way. 

I want to get the MongoDB data source updated to work with newer versions of MongoDB, and with MongoDB Atlas. Let me know if there is an interest and I can expedite this.

**Reply by wimjongman** [reply]
MongoDB support would be fantastic!

**Reply by TheLaserNinja** [reply]
So, we have a couple of ways of doing this now.
1: BI Connector for Atlas
2: SQL Connector (which is replacing BI Connector)
3: MongoDB JDBC Driver

I confirmed BI Connector works currently using the "MySQL JDBC Driver". Works fine. I haven't checked out SQL Connector, it is still pre-release. What I have been focusing on is 2 things, the MongoDB JDBC Driver and getting the out of the box MongoDB driver that we currently have working. There is work to be done on the MongoDB JDBC Driver, it appears to be missing some dependencies and the JDBC driver loader in BIRT doesn't find the class. I'll figure out what is going on there. I suspect it is on the driver side, so I'll either patch it there, or patch it on the BIRT side if necessary. For the out of the box one we currently have, I believe MongoDB changed so many things with the merging of WiredTiger and change of the transport protocol that the ODA we currently have will need some re-working. I will reach out to the product manager for Mongo's drivers project, and see if we can't just drop the MongoDB Java driver in so we have all the native features in Mongo, such as time-series collection support, Atlas support, etc.
